core/src/main/java/io/ably/lib/realtime/ConnectionRecoveryKey.java

```diff
@@ -0,0 +1,34 @@
+package io.ably.lib.realtime;
+
+import com.google.gson.JsonSyntaxException;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import io.ably.lib.util.Serialisation;
+
+public class ConnectionRecoveryKey {
+
+    public String connectionKey;
+    public long msgSerial;
```

Comment:
Can you make those private and create getters for them if needed?

Reply:
I don't think we should repeat the patterns before and we should try to use the best practices as much as we can in new classes. As this is still a Java library I think we should follow best practices for it. So object state must always be exposed through a method - and they must be immutable if possible. 
I don't understand what you mean by code clutter. Exposing object states as public goes against the principles of encapsulation

Reply:
As this is a simple data object it is not necessary to preserve encapsulation. 
Best practices change over time, kotlin best practices differ from java ones and even though we are not in kotlin yet we will move towards it which will make all those changes obsolete.
Generating all getters, setters, and map manipulation functions will make clutter in this class.
I can implement your suggested changes if you strongly agree with them.

Reply:
> As this is a simple data object it is not necessary to preserve encapsulation.

Can this object be accessed from outside and state be manipulated by multiple callers and/or threads? What's the difference between this being a data object or another object from the callers' perspective? You still expose state that needs to be encapsulated and can be modified by different actors putting it in an inconsistent state. I don't think we can make an exception for this.

> Best practices change over time, kotlin best practices differ from java ones and even though we are not in kotlin yet we will move towards it which will make all those changes obsolete.

I do not think Kotlin best practices differ that much in terms of object encapsulation and immutability. In fact Kotlin best practices recommends more immutability and state protection. `val` is exactly for this. The shorthand accessing object states is not equivalent to defining public mutable states like in here.

>  Generating all getters, setters, and map manipulation functions will make clutter in this class.

Yes, but you cannot really exchange code safety with uncluttered code. Plus, do you really need setters  here? Could you have defined one or more of these as `final` to make state immutable and only assign state when creating the object?

> I can implement your suggested changes if you strongly agree with them.

As above, yes please

Reply:
I do not think this can be accessed outside and manipulated by different threads at the same time as this is used on initializing ably realtime and when we want to save external recovery key and that is a getter.
I have implemented suggested changes here https://github.com/ably/ably-java/pull/842/commits/18415fc1b0351392c71e6b41f3d23d055dc3ff63

Reply:
> I do not think this can be accessed outside and manipulated by different threads 

Of course it can. That's what `public` keyword [is for](https://en.wikibooks.org/wiki/Java_Programming/Keywords/public). 

> as this is used on initializing ably realtime and when we want to save external recovery key and that is a getter.

That's the design intention behind it. The way it was designed make the use of this class much more beyond this use case.

Reply:
Thanks looks much better now!

Reply:
I do not think end user can use our class in the way it would compromise SDK. End user can use it to create new instance of it but no way to pass it in our sdk.
Any way, it is good this way as well so I will close this conversation.

Reply:
> I do not think end user can use our class in the way it would compromise SDK. End user can use it to create new instance of it but no way to pass it in our sdk.

This is not only a matter of exposing things to the end user. Exposing it to the SDK is always a risk and design flaw